samples/template/src/main.c

```diff
@@ -0,0 +1,517 @@
+		break;
+	case APP_EVENT_NOT_READY:
+		smf_set_state(SMF_CTX(&ctx->sm), &app_states[STATE_SIDEWALK_NOT_READY]);
+		break;
+	case APP_EVENT_SEND:
+		e = sid_put_msg(ctx->handle, &app_send_buffer.msg, &app_send_buffer.desc);
+		if (e) {
+			LOG_ERR("sid send err %d", (int)e);
+		}
+		LOG_INF("sid send (type: %d, id: %u)", (int)app_send_buffer.desc.type,
+			app_send_buffer.desc.id);
+		break;
+	default:
+		LOG_WRN("event: unknow %d", ctx->sm.event);
+		break;
+	}
+}
+
+static void button_changed(uint32_t button_state, uint32_t has_changed)
+{
+	uint32_t buttons = button_state & has_changed;
+	int err = 0;
+
+	if (buttons & DK_BTN1_MSK) {
+		LOG_INF("button 1");
+
+		static char payload[] = "hello";
+		app_send_buffer = (struct app_msg){
+			.msg = {
+				.data = payload,
+				.size = sizeof(payload),
+			},
+			.desc = {
+				.type = SID_MSG_TYPE_NOTIFY,
+				.link_type = SID_LINK_TYPE_ANY,
+				.link_mode = SID_LINK_MODE_CLOUD,
+			},
+		};
+		err = app_event_send(APP_EVENT_SEND);
+	}
+	if (buttons & DK_BTN2_MSK) {
+		LOG_INF("button 2");
+		err = app_event_send(APP_EVENT_LINK_SWITCH);
+	}
+	if (buttons & DK_BTN3_MSK) {
+		LOG_INF("button 3");
+	}
+	if (buttons & DK_BTN4_MSK) {
+		LOG_INF("button 4");
+		err = app_event_send(APP_EVENT_FACTORY_RESET);
+	}
+
+	if (err) {
+		LOG_ERR("send event err %d", err);
+	};
+}
+
+static int app_event_send(app_event_t event)
+{
+	return k_msgq_put(&sid_ctx.sm.msgq, (void *)&event, K_NO_WAIT);
+}
+
+static void app_thread_entry(void *context, void *unused, void *unused2)
+{
+	LOG_INF("App");
+
+	ARG_UNUSED(unused);
+	ARG_UNUSED(unused2);
+
+	struct sid_ctx_s *ctx = (struct sid_ctx_s *)context;
+
+	struct sid_event_callbacks event_callbacks = {
+		.context = ctx,
+		.on_event = on_sidewalk_event,
+		.on_msg_received = on_sidewalk_msg_received,
+		.on_msg_sent = on_sidewalk_msg_sent,
+		.on_send_error = on_sidewalk_send_error,
+		.on_status_changed = on_sidewalk_status_changed,
+		.on_factory_reset = on_sidewalk_factory_reset,
+	};
+
+	ctx->config = (struct sid_config){
+		.link_mask = SID_LINK_TYPE_1,
+		.callbacks = &event_callbacks,
+		.time_sync_periodicity_seconds = 7200,
+		.link_config = app_get_ble_config(),
+		.sub_ghz_link_config = app_get_sub_ghz_config(),
+	};
+
+	k_msgq_init(&ctx->sm.msgq, msgq_buffer, sizeof(app_event_t),
+		    CONFIG_SIDEWALK_THREAD_QUEUE_SIZE);
+	smf_set_initial(SMF_CTX(&ctx->sm), &app_states[STATE_INIT]);
+
+	while (1) {
+		LOG_INF("waiting for event...");
+		int err = k_msgq_get(&ctx->sm.msgq, &ctx->sm.event, K_FOREVER);
+		if (!err) {
+			if (smf_run_state(SMF_CTX(&ctx->sm))) {
+				break;
+			}
+		} else {
+			LOG_ERR("msgq err %d", err);
+		}
+	}
+
+	LOG_ERR("Application failed. You should never see this message.");
+}
+
+int main(void)
+{
+	LOG_INF("Main");
+
+	if (dk_buttons_init(button_changed)) {
+		LOG_ERR("Cannot init buttons");
+	}
+
+	if (dk_leds_init()) {
+		LOG_ERR("Cannot init leds");
+	}
+
+	(void)k_thread_create(&app_thread, app_thread_stack,
```

Comment:
If we spawn sidewalk thread here and the main thread dies, do we need to create new thread ? Maybe we could just call app_thread_entry directly ?  we would need to adjust the stack size and thread priority for main, but maybe it would be better than creating another thread ?

Reply:
Considering the fragility of timings, priorities, and amount of memory required by Sidewalk to operate (especially with subghz radio) I would recommend to keep in it's own thread.